e2e/internal/e2e/profile.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package e2e
+
+import (
+	"testing"
+
+	"github.com/sylabs/singularity/internal/pkg/fakeroot"
+	"github.com/sylabs/singularity/internal/pkg/test/tool/require"
+	"github.com/sylabs/singularity/internal/pkg/util/user"
+)
+
+// SingularityProfile represents a Singularity execution profile.
+type SingularityProfile uint8
+
+const (
+	// UserProfile is the execution profile with a regular user.
+	UserProfile SingularityProfile = iota + 1
+	// RootProfile is the execution profile with root.
+	RootProfile
+	// FakerootProfile is the execution profile with fakeroot.
+	FakerootProfile
+	// UserNamespaceProfile is the execution profile with a user namespace.
+	UserNamespaceProfile
+	// RootUserNamespaceProfile is the execution profile with root and a user namespace.
+	RootUserNamespaceProfile
+)
+
+// Profiles groups all Singularity execution profiles.
+var Profiles = []SingularityProfile{
+	UserProfile,
+	RootProfile,
+	FakerootProfile,
+	UserNamespaceProfile,
+	// NEED FIX: disabled until tests are fixed to run it correctly
+	// RootUserNamespaceProfile,
+}
+
+// CheckRequirements checks and ensures that the corresponding
+// execution profile has all the requirements, the current test
+// is skipped if not.
+func (p SingularityProfile) CheckRequirements(t *testing.T) {
+	switch p {
+	case UserNamespaceProfile, RootUserNamespaceProfile:
+		require.UserNamespace(t)
+	case FakerootProfile:
+		require.UserNamespace(t)
+		// now check that current user has valid mappings
+		// in /etc/subuid and /etc/subgid
+		if _, err := fakeroot.GetIDRange(fakeroot.SubUIDFile, uint32(origUID)); err != nil {
+			t.Skipf("fakeroot configuration error: %s", err)
+		}
+		if _, err := fakeroot.GetIDRange(fakeroot.SubGIDFile, uint32(origUID)); err != nil {
+			t.Skipf("fakeroot configuration error: %s", err)
+		}
+	}
+}
+
+// Match returns if one the provided profiles match the current one.
+func (p SingularityProfile) Match(profiles ...SingularityProfile) bool {
+	for _, profile := range profiles {
+		if profile == p {
+			return true
+		}
+	}
+	return false
+}
+
+// Privileged returns if the profile requires to run with
+// privileges or not.
+func (p SingularityProfile) Privileged() bool {
+	switch p {
+	case RootProfile, RootUserNamespaceProfile:
+		return true
+	}
+	return false
+}
+
+// WithArgs returns Singularity arguments for the corresponding
+// profile if any.
+func (p SingularityProfile) WithArgs(args []string) []string {
+	switch p {
+	case FakerootProfile:
+		return append([]string{"--fakeroot"}, args...)
```

Comment:
these functions here (the previous one, this one, the next one) are what makes me think the code could be implemented differently, but I don't understand the use case enough to figure out of there's an advantage to this or not.

If you had:

```
type Profile interface {
    Privileged() bool
    Args() []string
    User() (*user.User, error)
}
```

profiles would be types implementing the interface, so for example:

```
type RootProfile struct{}

func (p RootProfile) Privileged() bool {
    return true
}

func (p RootProfile) Args() []string {
    return nil
}

func (p RootProfile) User() (*user.User, error) {
    return user.GetPwUID(0)
}
```

it's a little more typing, but the implementations are much more straightforward.

Whether this makes sense or not hinges on what the purpose of Match is.

Reply:
The idea is to have static and not alterable profiles, because there is nothing to set dynamically, so we can avoid allocation step